proof/refine/ARM_HYP/Tcb_R.thy

```diff
@@ -1920,19 +1976,31 @@ lemma setTLSBase_invs'[wp]:
    \<lbrace>\<lambda>rv. invs'\<rbrace>"
   by (wpsimp simp: invokeTCB_def)
 
-lemma threadSet_flags_invs[wp]:
-  "threadSet (tcbFlags_update f) t \<lbrace>invs'\<rbrace>"
-  by (wpsimp wp: threadSet_invs_trivial)
+lemma threadSet_flags_invs'[wp]:
+  "\<lbrace>invs' and K (\<forall>flags. flags \<le> mask numFlags \<longrightarrow> f flags \<le> mask numFlags)\<rbrace>
+   threadSet (tcbFlags_update f) t
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  by (rule hoare_gen_asm) (wpsimp wp: threadSet_invs_trivial)
+
+lemma setFlags_invs'[wp]:
+  "\<lbrace>invs' and K (flags \<le> mask numFlags)\<rbrace>
+   setFlags t flags
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  by (wpsimp simp: setFlags_def wp: threadSet_flags_invs')
```

Comment:
I'm curious why this requires the precondition... from discussions, it seemed that extra flags get ignored. Is this a constraint resulting from an update to TCB validity? I understand needing the precondition for [c]corres, but confused why for valid.

Reply:
Does adding a comment make sense?

Reply:
I don't think so, I think it makes more sense in context when you can see that `setFlags` is just a wrapper around `threadSet`, and that this lemma is just a simplification of `threadSet_flags_invs'`. Most other `threadSet` lemmas have a similar precondition constraining the value being set so that TCB validity is maintained.

Reply:
I'm convinced, fine to leave as is.